org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/folding/IFoldingRangeAcceptor.java

```diff
@@ -0,0 +1,27 @@
+/*******************************************************************************
+ * Copyright (c) 2010 Michael Clay and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package org.eclipse.xtext.ide.editor.folding;
+
+/**
+ * @author Michael Clay - Initial contribution and API
+ * @author Sebastian Zarnekow - Distinguish between total and identifying region
+ * @author Mark Sujew - Ported to IDE project
+ * 
+ * @since 2.25
+ */
+public interface IFoldingRangeAcceptor<T> {
+
+	void accept(int offset, int length);
+	void accept(int offset, int length, String kind);
+	void accept(int offset, int length, boolean initiallyFolded);
+	void accept(int offset, int length, T addedParams);
```

Comment:
What is the purpose of the type parameter T?

Reply:
The purpose is to keep compatibility with the original folding implementation of [xtext-eclipse](https://github.com/eclipse/xtext-eclipse/tree/master/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding), which as well contains a type parameter.

Reply:
In Eclipse, a folding region can carry information about the text that is to be shown if that region is collapsed. This capability does not seem to be present in LSP.
OTOH the LSP protocol has a notion of a `kind` which is not at all relevant in Eclipse.
Am I right in the assumption that you are implementing the superset of both frontend's capabilities? Would it make sense to use ITextRegion directly instead of the type parameter? It would also allow to give it a better name compared to `addedParams`

Reply:
Yes, implementing a superset of both capabilities was the idea behind that. I thought about using ITextRegion as well. However, I tried not to break compatibility for future would-be implementors of the new API that didn't use ITextRegion in the past.

I thought about the intersection based solution as well. However, I found it to be quite bothersome to build acceptors for each frontend solution. Alternatively, an `IFoldingRangeAcceptor` could accept any FoldingRange implementation, instead of accepting the individual parts of a FoldingRange. What do you think of this?

Reply:
I've a weak tendency towards an *explicit* superset of both known worlds, too.

In Eclipse, we have the `org.eclipse.jface.text.source.projection.IProjectionPosition` and in LSP the `FoldingRange`.
Xtext simplifies the IProjectionPosition by chosing exactly one text region that is supposed to stay whereas the are before and after that region is the part of the document that will not be visible if the range is collapsed.
Apart from that, Xtext offers a flag `initiallyCollapsed`.
LSP does not offer these but has a notion of a `kind` instead.
So our superset could be modelled as `offset, length, initiallyFolded, kind, visualPlaceholderRegion(=ITextRegion)`. A little annoying since the intersection of both frontends is only the tuple of `offset, length`. 

Let's assume we use `FoldingRange` in the acceptor. We'd still want to be able to implement folding only once and benefit the most on both platforms. So FoldingRange would support the superset. Essentially we'd replace the accept(..) methods by overloaded constructors or factory methods. That would only add another layer of indirection. It'd render the acceptor obsolete since the API could be based on a SortedSet directly. The disadvantage: The acceptor pattern allows to create a pseudo-smart acceptor, e.g. one that will only accept regions larger than X, that will drop regions after n other regions have already been registered or that will even do cancellation checks internally and drop out automatically. We probably want to stick to the acceptor based solution with the plenty of overloaded accept methods. Maybe it's sufficient to add reasonable JavaDoc about the limitations of both platforms on the fully fledged overload of accept? And dropping the type parameter in favour of a concrete ITextRegion should be possible if I'm not mistaken.

Reply:
Ok, so I now implemented everything as a superset. Sadly, I couldn't reproduce the exact behavior of [this code](https://github.com/eclipse/xtext-eclipse/blob/39c7b77e5100dc15284cd30c20f20cdf15b110e5/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/DefaultFoldingRegionAcceptor.java#L43-L50), since I couldn't figure out how to get the length of a certain line without access to `IXtextDocument` or `Document`. Is it possible to figure this out via the NodeModel?

Reply:
@Darkgaja  IIUIC the code will alter invalid significant regions such that only the first line of the region is taken. So basically it'll start from the significant regions offset until it finds the first line-break or the end-offset of the significant region. It is possible to achieve that based on the root node's text along with NodeModelUtils.getLineAndColumn(INode, int) to compare the line numbers.

Reply:
Great, thanks for the info. I implemented it via the getLineAndColumn method. Could you take a look at it? I think I addressed every issue you had.